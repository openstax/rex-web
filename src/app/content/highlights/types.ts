import { Highlight, HighlightColorEnum, HighlightsSummary, NewHighlight } from '@openstax/highlighter/dist/api';
import { ScrollTarget } from '../../navigation/types';
import { FiltersChange } from '../components/popUp/types';
import { LinkedArchiveTreeNode } from '../types';

export interface SummaryHighlights {
  [locationId: string]: {[pageId: string]: HighlightData[]};
}

export type OrderedSummaryHighlights = Array<{
  location: LinkedArchiveTreeNode,
  pages: Array<{
    pageId: string;
    highlights: HighlightData[];
  }>
}>;

export interface SummaryFilters {
  locationIds: string[];
  colors: HighlightColorEnum[];
}

export interface SummaryFiltersUpdate {
  locations?: FiltersChange<LinkedArchiveTreeNode>;
  colors?: FiltersChange<HighlightColorEnum>;
}

export type CountsPerSource = NonNullable<HighlightsSummary['countsPerSource']>;
export type HighlightColorCounts = CountsPerSource[string];

export type SummaryHighlightsPagination = null | {
  sourceIds: string[];
  page: number;
  perPage: number;
};

export interface State {
  currentPage: {
    pageId: string | null,
    highlights: null | HighlightData[];
    hasUnsavedHighlight: boolean;
    focused?: string;
  };
  summary: {
    open: boolean,
    pagination: SummaryHighlightsPagination,
    totalCountsPerPage: CountsPerSource | null;
    filters: SummaryFilters,
    loading: boolean;
    highlights: SummaryHighlights | null;
  };
}

export interface HighlightScrollTarget extends ScrollTarget {
  type: 'highlight';
  id: string;
}

// scopeId should not be nullable in the swagger
export type NewHighlightPayload = NewHighlight & {scopeId: string};
export type HighlightData = Highlight & {scopeId: string};
