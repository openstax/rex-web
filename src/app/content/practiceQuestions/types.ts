import { LinkedArchiveTreeSection } from '../types';

export interface State {
  summary: PracticeQuestionsSummary | null;
  selectedSection: LinkedArchiveTreeSection | null;
  currentQuestionIndex: number | null;
  questions: PracticeQuestion[];
  questionAnswers: QuestionAnswers;
  loading: boolean;
}

export interface PracticeQuestionsSummary {
  countsPerSource: {
    [key: string]: number;
  };
}

export interface PracticeAnswer {
  id: number;
  content_html: string;
  correctness: '0.0' | '1.0';
  feedback_html: string;
}

export interface PracticeQuestion {
  group_uuid: string;
  uid: string;
  stem_html: string;
  answers: PracticeAnswer[];
}

export type PracticeQuestions = PracticeQuestion[];

export interface QuestionAnswers {
  [key: string]: PracticeAnswer | null;
}

export interface SetPracticeAnswer {
  questionId: string;
  answer: PracticeAnswer | null;
}
