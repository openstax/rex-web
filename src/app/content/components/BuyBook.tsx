import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { useSelector } from 'react-redux';
import styled from 'styled-components/macro';
import { textRegularSize, textRegularStyle } from '../../components/Typography';
import theme from '../../theme';
import { buyPrintConfig } from '../selectors';
import { contentTextWidth } from './constants';
import { disablePrint } from './utils/disablePrint';

// tslint:disable-next-line:variable-name
const BuyBookAlignment = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  max-width: ${contentTextWidth}rem;
  overflow: visible;
  ${disablePrint}
`;

// tslint:disable-next-line:variable-name
const BuyBookLink = styled.a`
  ${textRegularSize};
  font-size: 1.9rem;
  display: flex;
  align-items: center;
  justify-content: center;
  text-decoration: none;
  height: 5rem;
  width: 19rem;
  color: ${theme.color.primary.orange.base};
  border: solid 0.1rem;
  font-weight: 700;
`;

// tslint:disable-next-line:variable-name
const BuyPrintDisclosure = styled.p`
  flex: 1;
  align-self: stretch;
  margin: 1.6rem 0 0 0;
  ${textRegularStyle}
  font-size: 1.2rem;
  line-height: 1.7rem;
`;

// tslint:disable-next-line: variable-name
const BuyBook = () => {
  const config = useSelector(buyPrintConfig);
  const formatMessage = useIntl().formatMessage;

  if (!config) { return null; }

  return <BuyBookAlignment>
    <BuyBookLink
      aria-label={formatMessage({id: 'i18n:toolbar:buy-book:aria-label:text'})}
      target='_blank'
      rel='noopener'
      href={config.url}
      data-analytics-label='buy-book'
    >
      <FormattedMessage id='i18n:toolbar:buy-book:text'>
        {(msg) => msg}
      </FormattedMessage>
    </BuyBookLink>
    {config.disclosure && <BuyPrintDisclosure>{config.disclosure}</BuyPrintDisclosure>}
  </BuyBookAlignment>;
};

export default BuyBook;
