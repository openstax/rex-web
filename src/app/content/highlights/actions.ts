import { Highlight, NewHighlight, UpdateHighlightRequest } from '@openstax/highlighter/dist/api';
import { createStandardAction } from 'typesafe-actions';
import { LocationFilters } from '../components/popUp/types';
import {
  CountsPerSource,
  HighlightData,
  SummaryFilters,
  SummaryFiltersUpdate,
  SummaryHighlights,
  SummaryHighlightsPagination,
} from './types';

export const focusHighlight = createStandardAction('Content/Highlights/focus')<string>();
export const clearFocusedHighlight = createStandardAction('Content/Highlights/clear')();
export const createHighlight = createStandardAction('Content/Highlights/create')<NewHighlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const requestDeleteHighlight = createStandardAction('Content/Highlights/requestDelete')<Highlight, {
  locationFilterId: string,
  pageId: string,
}>();
export const receiveDeleteHighlight = createStandardAction('Content/Highlights/receiveDelete')<Highlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const updateHighlight = createStandardAction('Content/Highlights/update')<UpdateHighlightRequest, {
  revertingAfterFailure?: boolean,
  preUpdateData: UpdateHighlightRequest,
  locationFilterId: string,
  pageId: string,
}>();
export const receiveHighlights = createStandardAction(
  'Content/Highlights/receive'
)<{highlights: HighlightData[], pageId: string}>();
export const setAnnotationChangesPending = createStandardAction('Content/Highlights/setAnnotationChangesPending')<
  boolean
>();

export const openMyHighlights = createStandardAction('Content/Highlights/Summary/open')<void>();
export const closeMyHighlights = createStandardAction('Content/Highlights/Summary/close')<void>();
export const initializeMyHighlightsSummary = createStandardAction('Content/Highlights/Summary/init')<void>();

export const printSummaryHighlights = createStandardAction('Content/Highlights/Summary/print')();
export const toggleSummaryHighlightsLoading = createStandardAction('Content/Highlights/Summary/loading')<boolean>();

export const loadMoreSummaryHighlights = createStandardAction('Content/Highlights/Summary/loadMore')();
export const setSummaryFilters = createStandardAction('Content/Highlights/Summary/setFilters')<
  Partial<SummaryFilters>
>();
export const updateSummaryFilters = createStandardAction('Content/Highlights/Summary/updateFilters')<
  Partial<SummaryFiltersUpdate>
>();
export const receiveSummaryHighlights = createStandardAction('Content/Highlights/Summary/receiveHighlights')<
  SummaryHighlights,
  {
    pagination: SummaryHighlightsPagination,
    filters?: SummaryFilters,
    isStillLoading?: boolean
  }
>();
export const receiveHighlightsTotalCounts = createStandardAction(
  'Content/receiveHighlightsTotalCounts'
)<CountsPerSource, LocationFilters>();
