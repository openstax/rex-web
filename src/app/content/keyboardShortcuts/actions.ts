import { createStandardAction } from 'typesafe-actions';

export const openKeyboardShortcutsMenu = createStandardAction('Content/KeyboardShortcuts/Menu/open')<void>();
export const closeKeyboardShortcutsMenu = createStandardAction('Content/KeyboardShortcuts/Menu/close')<void>();
