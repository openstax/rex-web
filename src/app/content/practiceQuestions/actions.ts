import { createStandardAction } from 'typesafe-actions';
import { LinkedArchiveTreeSection } from '../types';
import { PracticeQuestion, PracticeQuestionsSummary, SetPracticeAnswer } from './types';

export const setSelectedSection = createStandardAction(
  'Content/PracticeQuestions/setSelectedSection'
)<LinkedArchiveTreeSection | null>();
export const setQuestions = createStandardAction('Content/PracticeQuestions/setQuestions')<PracticeQuestion[]>();
export const nextQuestion = createStandardAction('Content/PracticeQuestions/nextQuestion')<void>();
export const setAnswer = createStandardAction('Content/PracticeQuestions/setAnswer')<SetPracticeAnswer>();
export const finishQuestions = createStandardAction('Content/PracticeQuestions/finishQuestions')<void>();

export const receivePracticeQuestionsSummary = createStandardAction(
  'Content/PracticeQuestions/Summary/receive'
)<PracticeQuestionsSummary>();

export const openPracticeQuestions = createStandardAction('Content/PracticeQuestions/open')<void>();
export const closePracticeQuestions = createStandardAction('Content/PracticeQuestions/close')<void>();
