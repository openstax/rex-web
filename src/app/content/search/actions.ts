import { createAction, createStandardAction } from 'typesafe-actions';
import { SearchScrollTarget, SelectedResult, State } from './types';

export const receiveSearchResults = createAction('Content/Search/receiveResults', (action) =>
  (
    results: Exclude<State['results'], null>,
    meta?: {searchScrollTarget?: SearchScrollTarget}
  ) => action(results, meta)
);

export const requestSearch = createAction('Content/Search/request', (action) =>
  (
    query: string,
    meta?: {
      isResultReload: boolean,
      searchScrollTarget?: SearchScrollTarget,
    }
  ) => action(query, meta)
);

export const clearSearch = createStandardAction('Content/Search/clear')();
export const openMobileToolbar = createStandardAction('Content/Search/openMobileToolbar')();
export const openSearchResultsMobile = createStandardAction('Content/Search/open')();
export const closeSearchResultsMobile = createStandardAction('Content/Search/close')();

export const selectSearchResult = createStandardAction('Content/Search/selectResult')<SelectedResult>();

export const openSearchInSidebar = createStandardAction('Content/Search/openSearchInSidebar')();
