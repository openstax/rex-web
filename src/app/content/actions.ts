import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as practiceQuestionsActions from './practiceQuestions/actions';
import * as searchActions from './search/actions';
import * as studyGuidesActions from './studyGuides/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;
export const studyGuides = studyGuidesActions;
export const practiceQuestions = practiceQuestionsActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const receiveBuyPrintConfig = createStandardAction('Content/receiveBuyPrintConfig')<State['buyPrint']>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();
export const receivePageNotFoundId = createStandardAction('Content/receivePageNotFoundId')<string>();

export const openNudgeStudyTools = createStandardAction('Content/openNudgeStudyTools')();
export const closeNudgeStudyTools = createStandardAction('Content/closeNudgeStudyTools')();
