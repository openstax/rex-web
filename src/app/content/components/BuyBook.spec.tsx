import React from 'react';
import renderer from 'react-test-renderer';
import createTestStore from '../../../test/createTestStore';
import TestContainer from '../../../test/TestContainer';
import { Store } from '../../types';
import { receiveBuyPrintConfig } from '../actions';
import BuyBook from './BuyBook';

describe('BuyBook', () => {
  let store: Store;

  beforeEach(() => {
    store = createTestStore();
  });

  it('renders when config is available', () => {
    store.dispatch(receiveBuyPrintConfig({url: 'https://example.com', disclosure: 'asdf'}));
    const component = renderer.create(<TestContainer store={store}>
      <BuyBook />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('returns null', () => {
    const component = renderer.create(<TestContainer store={store}>
      <BuyBook />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
