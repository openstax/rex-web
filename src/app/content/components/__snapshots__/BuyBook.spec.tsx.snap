// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuyBook renders when config is available 1`] = `
<div
  className="BuyBook__BuyBookAlignment-s2exzh-0 dsrSOQ"
>
  <a
    aria-label="Buy book"
    className="BuyBook__BuyBookLink-s2exzh-1 hNBOXB"
    data-analytics-label="buy-book"
    href="https://example.com"
    rel="noopener"
    target="_blank"
  >
    Order a print copy
  </a>
  <p
    className="BuyBook__BuyPrintDisclosure-s2exzh-2 dPPdHn"
  >
    asdf
  </p>
</div>
`;

exports[`BuyBook returns null 1`] = `null`;
